src/library/QueryBuilder/QueryBuilder.ts

```diff
@@ -0,0 +1,118 @@
+  readonly EQUALS = "=";
+  readonly COLON = ":";
+  readonly QUERY_DELIMETER = "&";
+  readonly COMMA = ",";
+  private baseResource: String = "";
+  private singularQueries: String[] = [];
+  private sortQueries: String[] =  [];
+
+  /**
+   * @returns base resource for the query
+   */
+  getBaseResource(): String {
+    return this.baseResource;
+  }
+
+  /**
+   * @param resourceType Base ResourceType for the query 
+   * sets base resource type for the query
+   * i.e. which resource the query will performed for
+   */
+  setBaseResource(resourceType: String) {
+    this.baseResource = resourceType;
+    return this;
+  }
+
+  /**
+   * @param searchParameter SearchParameter reference to targeted resource
+   */
+  include(searchParameter: String) {
+    this.singularQueries.push(this.INCLUDE_KEYWORD + this.EQUALS + encodeURIComponent(this.baseResource + this.COLON + searchParameter));
+    return this;
+  }
+
+  /**
+   * @param resourceType Source ResourceType 
+   * @param searchParameter SearchParameter reference from target resource to baseResource
+   */
+  revinclude(resourceType: String, searchParameter: String) {
+    this.singularQueries.push(this.REV_INCLUDE_KEYWORD + this.EQUALS +  encodeURIComponent(resourceType + this.COLON + searchParameter));
+    return this;
+  }
+
+  /**
+   * _include all references in query
+   */
+  includeAll() {
+    this.singularQueries.push(this.INCLUDE_KEYWORD + this.EQUALS + this.WILDCARD_ASTERIK);
+    return this;
+  }
+
+  /**
+   * _revinclude all references in query
+   */
+  revincludeAll() {
+    this.singularQueries.push(this.REV_INCLUDE_KEYWORD + this.EQUALS + this.WILDCARD_ASTERIK);
+    return this;
+  }
+
+  /**
+   * @param searchParameter search parameter for the element to sort on
+   * @param sortOrder ASCENDING or DESCENDING
+   */
+  sort(searchParameter: String, sortOrder: SORT_ORDER) {
+    if(sortOrder === SORT_ORDER.ASCENDING) {
+      this.sortQueries.push(searchParameter);
+    } else {
+      this.sortQueries.push("-" + searchParameter);
+    }
+    return this;
+  }
+
+  /**
+   * Resets queries to it empty state
+   */
+  resetQuery() {
+    this.sortQueries = [];
+    this.singularQueries = [];
+    this.baseResource = "";
+    return this;
+  }
+
+  /**
+   * @returns complete generated query with encoded parameter values
+   */
+  getCompleteUrl(): String {
+    let completeUrl = this.baseResource + "?";
+    if(this.singularQueries.length > 0) {
+      completeUrl +=  this.singularQueries.join(this.QUERY_DELIMETER);
+    }
+    completeUrl += this.createSortQuery();
+    return completeUrl; 
+  }
+
+  /**
+   * @returns complete generated query with decoded parameter values
+   */
+  getCompleteUrlDecoded(): String {
+    return decodeURIComponent(this.getCompleteUrl().toString()); 
+  }
+  
+  private createSortQuery(): String {
+    let sortQuery = "";
+    const sortDelimeter = this.singularQueries.length > 0 ? this.QUERY_DELIMETER : "";
```

Comment:
I don't like that this createSortQuery() function has to know about the field singularQueries. Refactoring will make this reference hard to trace. Either pass it in as a parameter or, better yet, put the logic for the delimiter in the calling function (getCompleteUrl()).

Reply:
i actually passed it in as a parameter just to keep the code a bit tidy.